### PR TITLE
feat: add ProveClosest method to SMT and SMST

### PR DESCRIPTION
**Pull request by h5law**
### Summary generated by Reviewpad on 25 Sep 23 17:51 UTC
This pull request includes the following changes:

- The file diff for file `example.go` includes the addition of a TODO comment to research whether the SiblingData is required and remove it if not.

- The file diff for file `smst.go` includes the addition of a new method called ProveClosest in the SMST struct. This method generates a SparseMerkleProof of inclusion for a given key, using the most common bits as the provided path. It returns the closest path, closest value hash, closest sum, the proof, and any error encountered during the process. Additionally, the Prove and Commit methods remain unchanged in this file.

- The file diff for file `types.go` includes the addition of two new methods ProveClosest in the SparseMerkleTree and SparseMerkleSumTree interfaces. These methods compute a Merkle proof of inclusion for a key in the tree that is closest to the provided path. The search for the closest key is based on the longest common prefix and the key with the most common bits as the provided path.

- The file diff for file `hash.go` includes the addition of an import statement for the "hash" package and a new function named "NoPrehashSpec" which returns a new TreeSpec with a nil Value Hasher and a nil Path Hasher. This function should only be used when values are already hashed and a path is used instead of a key during proof verification.

- The file diff for file `smt.go` includes the addition of a new method ProveClosest in the SMST struct, which generates a SparseMerkleProof of inclusion for the first key with the most common bits as the path provided. It also includes updates to the Prove method to return a proof object, the addition of the recursiveLoad method to recursively load a hash, the addition of the split method to split an extension node, and the addition of the expand method to return the inner node that represents the start of the singly linked list that an extension node represents.

- The file diff for file `smst_proofs_test.go` includes the addition of new test functions: TestSMST_ProveClosest, TestSMST_ProveClosestEmptyAndOneNode. These test the closest value proof in a sparse Merkle sum tree and test the closest value proof with an empty tree and a tree with a single node, respectively. Various updates and modifications were also made to the existing code and test cases in this file.

- The file diff for file `utils.go` includes the addition of a new type nilPathHasher with methods for a path and path size, a new function newNilPathHasher, a comment indicating that the GetPathBit function should be unexported, and a new function flipPathBit to flip the bit at a given position in the data. The setPathBit function was also updated to use bitwise XOR for flipping the bit at a given position. No changes were made to the countSetBits function.

- The file diff for file `smt_proofs_test.go` includes updates to the test cases in this file. A new test case TestSMT_ProveClosest was added to validate the closest key and value hash for a given path in the sparse Merkle tree. Another test case TestSMT_ProveClosestEmptyAndOneNode was added to test the ProveClosest function when the tree is empty or contains only one node. Some helper functions and variables were also added to support the new test cases.

Please review these changes in the code.

## Issue

Fixes N/A

Part of the work related to relay mining: paper can be viewed [here](https://arxiv.org/pdf/2305.10672.pdf)

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Adds ProveClosest method to SMT
- Adds ProveClosest method to SMST
- Adds unit tests

## Testing

- [x] **Task specific tests or benchmarks**: `go test ...`
- [x] **New tests or benchmarks**: `go test ...`
- [x] **All tests**: `go test -v`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling

### If Applicable Checklist

- [ ] Update any relevant README(s)
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Olshansk** [comment]
@h5law I'm going to review this this week. an you merge it with main?

I wanted to know if you still have your visualizer handy or accessible? Could be good for a visual check.

File: smt.go [hunk 1]
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+	err error, // the error value encountered
+) {
+	var siblings []treeNode
+	var sib treeNode
+	var parent treeNode
+	var parentDepth int
+
+	node := smt.tree
+	depth := 0
+	for depth < smt.depth() {
```
Comment (Olshansk) [inline smt.go:394]: Request: can you add some comments throughout the business logic? 

You have this in the lower half of the function.

For example, above the `if node == nil {`  could be a good place, amongst others.

File: smt.go [hunk 2]
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+		node, err = smt.resolveLazy(node)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if node == nil && depth == 0 {
+			break // DISCUSS_IN_THIS_PR: I dont think this is even possible
```
Comment (Olshansk) [inline smt.go:400]: This can only happen if the root is nil.

Outside of the forloop, you can check `if smt.tree == nil {return err}`

File: smt.go [hunk 3]
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+		}
+		parentDepth = depth
```
Comment (Olshansk) [inline smt.go:402]: We only use these inside the if structure, so should we assign them inside the if as well?

File: smt.go [hunk 4]
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+			parent, err = smt.resolveLazy(parent)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			inner, ok := parent.(*innerNode)
+			if !ok {
+				panic("parent not inner node")
+			}
+			if len(siblings) > 0 {
+				siblings = siblings[:len(siblings)-1]
```
Comment (Olshansk) [inline smt.go:414]: What are we trimming here?

File: smt.go [hunk 5]
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+			}
+			depth = parentDepth
+			if GetPathBit(path, depth) == left {
+				node = inner.rightChild
+			} else {
+				node = inner.leftChild
+			}
+		}
+		if _, ok := node.(*leafNode); ok {
```
Comment (Olshansk) [inline smt.go:423]:
```suggestion
		// End traversal once we have reached a leaf node
		if _, ok := node.(*leafNode); ok {
```

File: types.go [hunk 6]
```diff
@@ -30,6 +30,9 @@ type SparseMerkleTree interface {
 	Root() []byte
 	// Prove computes a Merkle proof of membership or non-membership of a key.
 	Prove(key []byte) (*SparseMerkleProof, error)
+	// ProveClosest computes a Merkle proof of inclusion for a key in the tree which is
+	// closest to the path provided
```
Comment (Olshansk) [inline types.go:34]: Add details on what closest means.

E.g. Most overlap in the path (i.e. # of bits) + default to left?

File: smt.go [hunk 7]
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+	}
+
+	// Retrieve the closest path and value hash if found
+	if node != nil {
+		leaf := node.(*leafNode)
+		closestPath, closestValueHash = leaf.path, leaf.valueHash
+	}
+	// Hash siblings from bottom up.
+	var sideNodes [][]byte
+	for i := range siblings {
```
Comment (Olshansk) [inline smt.go:460]: Why not `i, sibling := range siblings`?

File: smt_proofs_test.go [hunk 8]
```diff
@@ -143,3 +143,60 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
...
+	closestKey, closestValueHash, proof, err = smt.ProveClosest(path)
+	require.NoError(t, err)
+	require.NotEqual(t, proof, &SparseMerkleProof{})
+
+	result = VerifyProof(proof, root, closestKey, closestValueHash, NoPrehashSpec(sha256.New(), false))
+	require.True(t, result)
+	closestPath := sha256.Sum256([]byte("jackfruit"))
+	require.Equal(t, closestPath[:], closestKey)
+
+	// path = sha256.Sum256([]byte("xwordA188wordB110")) change 31st byte
```
Comment (Olshansk) [inline smt_proofs_test.go:181]: Can we do this automatically? Converting xwordA188wordB110 to a byte slice and then `res[30]...`?

File: smt_proofs_test.go [hunk 9]
```diff
@@ -143,3 +143,60 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
...
+	path = []byte{41, 6, 225, 86, 245, 213, 11, 141, 147, 82, 197, 13, 172, 115, 91, 244, 178, 217, 50, 38, 13, 171, 111, 56, 92, 209, 246, 148, 130, 113, 41, 171}
+	closestKey, closestValueHash, proof, err = smt.ProveClosest(path)
+	require.NoError(t, err)
+	require.NotEqual(t, proof, &SparseMerkleProof{})
+
+	result = VerifyProof(proof, root, closestKey, closestValueHash, NoPrehashSpec(sha256.New(), false))
+	require.True(t, result)
+	closestPath = sha256.Sum256([]byte("xwordA188wordB110"))
```
Comment (Olshansk) [inline smt_proofs_test.go:189]: NICE!

**Review by Olshansk — CHANGES_REQUESTED** [review]
@h5law Took a first review, but will also take a second one tomorrow.

I think this is one of the MOST critical pieces of v1, so would really appreciate a few more comments through the implementation.

File: smt.go [hunk 10]
```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
 	return proof, nil
 }
 
+// ProveClosest generates a SparseMerkleProof of inclusion for the key
+// with the most common bits as the path provided
```
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:381]: I am unsure what you mean here, it cannot bias a child as it follows a preset path, it biases the leaf with the longest common prefix (what we are looking for) but if along the path a `nil` node is hit (a placeholder) then it must have taken a wrong turn (we want a proof of inclusion). For example we took a right child and it was nil, so we go back to the parent and try the left child. Afterwards we continue to follow the path. This can happen either way and I dont see how it could be biased? LMK

File: smt.go (hunk 2 above)
Comment (h5law) [reply smt.go:400]: So this is actually caught in the very first resolve and error catch, if there is nothing to resolve we error out and we don't need this check at all.

File: smt.go (hunk 4 above)
Comment (h5law) [reply smt.go:414]: Added a comment, but if we hit this if statement we have hit a placeholder, prior to hitting it we added the sibling. This sibling is where we want to goto and thus should not be included in the side nodes as it would create an incorrect root hash. Ultimately this is just part of the backstepping.

File: smt.go (hunk 7 above)
Comment (h5law) [reply smt.go:460]: Because we want to hash the sidenodes from last to first in the slice there is no point taking the `sibling` from the range as we will not use it the next line shows how we get it using `i` and idk if there is a direct way to get this from range without sorting.

File: smt_proofs_test.go (hunk 8 above)
Comment (Olshansk) [reply smt_proofs_test.go:181]: @h5law Noticed that some of the comments haven't been addresses so I wanted to confirm if this is ready for re-review?

Comment (h5law) [reply smt_proofs_test.go:194]: I don't want to make this automatic simply because these are hardcoded it doesn't seem necessary. It would also require us checking the changed byte is not the same which just means extra logic.

LMK if you feel strongly the other way and I will do it.

File: options.go [hunk 11]
```diff
@@ -12,3 +14,15 @@ func WithPathHasher(ph PathHasher) Option {
 func WithValueHasher(vh ValueHasher) Option {
 	return func(ts *TreeSpec) { ts.vh = vh }
 }
+
+// NoPrehashSpec returns a new TreeSpec that has a nil Value and Path Hasher
+// NOTE: This should only be used when values are already hashed and a path is
```
[the remark this answers is not in the document]
Comment (h5law) [reply options.go:19]: Added some more context hope this clears it up

File: smt.go (hunk 3 above)
Comment (h5law) [reply smt.go:410]: I have updated these to actually work as intended I had missed that they were essentially no-op.

**Comment by h5law** [comment]
@Olshansk this PR should be good to go now ready for final reviews

File: smst_proofs_test.go [hunk 12]
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
...
+	require.NoError(t, err)
+	require.NotEqual(t, proof, &SparseMerkleProof{})
+
+	result = VerifySumProof(proof, root, closestKey, closestValueHash, closestSum, NoPrehashSpec(sha256.New(), true))
+	require.True(t, result)
+	closestPath := sha256.Sum256([]byte("testKey2"))
+	require.Equal(t, closestPath[:], closestKey)
+	require.Equal(t, closestSum, uint64(24))
+
+	// testValue4 is the neighbour of testValue2, by flipping the final bit of the
```
Comment (Olshansk) [inline smst_proofs_test.go:288]: Should these say `Value` or `Key`?

File: smst_proofs_test.go [hunk 13]
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
...
+	require.NoError(t, smst.Update([]byte("fob"), []byte("bof"), 18))
+	require.NoError(t, smst.Update([]byte("testKey"), []byte("testValue"), 21))
+	require.NoError(t, smst.Update([]byte("testKey2"), []byte("testValue2"), 24))
+	require.NoError(t, smst.Update([]byte("testKey3"), []byte("testValue3"), 27))
+	require.NoError(t, smst.Update([]byte("testKey4"), []byte("testValue4"), 30))
+
+	root = smst.Root()
+
+	// testKey2 is the child of an inner node which is the child of an extension node
+	// the extension node has the path bounds of [3, 7] by flipping these bits we force
```
Comment (Olshansk) [inline smst_proofs_test.go:272]: `path bounds of [3, 7] by flipping these`

1. I think this is missing some punctuation.
2. You mention the bounds [3, 7] here, and then have two separate bit flips: (3, 6) and (3, 7). Can you just clarify that since the extension has these bounds, we can run tests that flip any bits within those bounds?

File: smst_proofs_test.go [hunk 14]
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
...
+	flipPathBit(path[:], 3)
+	flipPathBit(path[:], 6)
+	closestKey, closestValueHash, closestSum, proof, err = smst.ProveClosest(path[:])
+	require.NoError(t, err)
+	require.NotEqual(t, proof, &SparseMerkleProof{})
+
+	result = VerifySumProof(proof, root, closestKey, closestValueHash, closestSum, NoPrehashSpec(sha256.New(), true))
+	require.True(t, result)
+	closestPath := sha256.Sum256([]byte("testKey2"))
+	require.Equal(t, closestPath[:], closestKey)
```
Comment (Olshansk) [inline smst_proofs_test.go:285]: Tbh, this is really fucking cool

File: smt_proofs_test.go [hunk 15]
```diff
@@ -155,3 +155,61 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
 	require.NoError(t, smn.Stop())
 	require.NoError(t, smv.Stop())
 }
+
+func TestSMT_ProveClosest(t *testing.T) {
```
Comment (Olshansk) [inline smt_proofs_test.go:159]: Please make the same updates here as you do in the other test

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: smt.go [hunk 16]
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+			parent = node
+		}
+		// resolve current node
+		node, err = smt.resolveLazy(node)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		// if we hit a nil node we backstep to the parent node and flip the path bit
+		// at the parent depth and select the other child
+		if node == nil {
```
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:418]: Not sure what you mean here making this `node != nil` makes a huge block within the if. currently its a single panic inside the branch as this should not happen

File: smt.go [hunk 17]
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+		} else {
+			depthDelta = 0
+		}
+		// end traversal when we hit a leaf node
+		if _, ok := node.(*leafNode); ok {
+			break
+		}
+		if ext, ok := node.(*extensionNode); ok {
+			length, match := ext.match(workingPath, depth)
+			if match {
```
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:439]: The thing is we cannot return or continue after this so both need to be in branches of the if/else as the result is we are either at a leaf or inner node afterwards and want to continue to the logic below. Personally I think swapping the shorter branch for the longer branch is purely stylistic here. LMK if you really think it helps to swap the branches around but I dont think it will help readability

File: smt.go [hunk 18]
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+				node = ext.child
+				node, err = smt.resolveLazy(node)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+			} else {
+				node = ext.expand()
+			}
+		}
+		inner := node.(*innerNode)
```
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:454]: There is only 3 types of nodes if it is neither a leaf or an extension node it must be an inner node

File: smt.go [hunk 19]
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+		siblings = append(siblings, sib)
+		depth += 1
+		depthDelta += 1
+	}
+
+	// Retrieve the closest path and value hash if found
+	if node == nil {
+		panic("no leaf node found")
+	}
+	leaf := node.(*leafNode)
```
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:469]: See comment above this is impossible to fail

File: smt.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:454]: @Olshansk this actually poses a good question, this is only possible to fail when the tree is empty. We are looking for a proof of inclusion and the updated return here results in a successful proof of inclusion of an empty tree. This could be checked user side (ie, is closestPath a placeholder is so this tree was empty) but I think the new logic here should be good.

File: smt.go [hunk 20]
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+			node, sib = inner.rightChild, inner.leftChild
+		}
+		siblings = append(siblings, sib)
+		depth += 1
+		depthDelta += 1
+	}
+
+	// Retrieve the closest path and value hash if found
+	if node == nil {
+		panic("no leaf node found")
```
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:467]: Returning an error here now see a comment above regarding empty trees

File: smt.go [hunk 21]
```diff
@@ -377,6 +377,119 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
...
+	for i := range siblings {
+		var sideNode []byte
+		sibling := siblings[len(siblings)-i-1]
+		sideNode = hashNode(smt.Spec(), sibling)
+		sideNodes = append(sideNodes, sideNode)
+	}
+	proof = &SparseMerkleProof{
+		SideNodes: sideNodes,
+	}
+	if sib != nil {
```
[the remark this answers is not in the document]
Comment (h5law) [reply smt.go:482]: So this seems to be left over from something I cant see it used anywhere in proof verification. Will leave it in for now and remove in another PR once I have fully gone through the code and checked it wont do any damage

File: smst_proofs_test.go [hunk 22]
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
...
+	smst = NewSparseMerkleSumTree(smn, sha256.New())
+
+	// insert random values
+	require.NoError(t, smst.Update([]byte("foo"), []byte("bar"), 5))
+	require.NoError(t, smst.Update([]byte("baz"), []byte("bin"), 5))
+	require.NoError(t, smst.Update([]byte("testKey"), []byte("testValue"), 5))
+	require.NoError(t, smst.Update([]byte("testKey2"), []byte("testValue"), 5))
+	require.NoError(t, smst.Update([]byte("testKey3"), []byte("testValue"), 5))
+	require.NoError(t, smst.Update([]byte("testKey4"), []byte("testValue"), 5))
+	// insert testing values that are similar
```
[the remark this answers is not in the document]
Comment (h5law) [reply smst_proofs_test.go:198]: I have removed these tests as the visual test is much better at doing what we are testing. These are simply covering the pre-existing tree traversal logic works which is the same as before

File: smst_proofs_test.go [hunk 23]
```diff
@@ -174,3 +174,129 @@ func TestSMST_ProofsSanityCheck(t *testing.T) {
...
+
+	result = VerifySumProof(proof, root, closestKey, closestValueHash, closestSum, NoPrehashSpec(sha256.New(), true))
+	require.True(t, result)
+	closestPath = sha256.Sum256([]byte("3xwordA250wordB7"))
+	require.Equal(t, closestPath[:], closestKey)
+	require.Equal(t, closestSum, uint64(11))
+}
+
+// ProveClosest test against a visual representation of the tree
+// See: https://i.imgur.com/cPJObIy.png
```
[the remark this answers is not in the document]
Comment (h5law) [reply smst_proofs_test.go:243]: How did you do this?

File: smst_proofs_test.go (hunk 13 above)
Comment (h5law) [reply smst_proofs_test.go:272]: 1. Unsure what you mean?
2. We can flip any bits in the 3-6 range and still get the result of testKey2 as these are in the path bounds (flipping other bits changes the path in a way that results in more than a backstep but we are forcing the proof here to do what we want)
	- so flipping bits: 3,4,5,6 result in testKey2
	- flipping bit 7 results in testKey4

File: utils.go [hunk 24]
```diff
@@ -26,6 +37,13 @@ func setPathBit(data []byte, position int) {
 	data[position/8] = byte(n)
 }
 
+// flipPathBit flips the bit at an offset from the most significant bit
+func flipPathBit(data []byte, position int) {
```
[the remark this answers is not in the document]
Comment (h5law) [reply utils.go:41]: I wanna keep it as `flipPathBit` to match `setPathBit` and `GetPathBit` (will be unexporting this now we dont need it in v1 anymore)

**Comment by h5law** [comment]
@Olshansk I addressed the comments and replied to some of the outstanding ones with reasonings, there are a few PRs that should come out of this one:
1. Look into removing `SiblingData` from proofs
2. Unexport `GetPathBit`
Otherwise I think the comments should clear things up that may have been unclear. I removed the original test cases in favour of the visual ones as what we are forcing is backstepping logic. Otherwise the proof functions the same and this logic is covered elsewhere.

File: smt.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply smt.go:421]: This is what I had in mind: 
```go
if node != nil {
  depthDelta = 0
} else {
 // large code block}
}
```

File: smt.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply smt.go:446]: This is what I had in mind:
```go
if !match {
   node = ext.expand()
} else {
  // large code block
}
```

I personally think it'll help readability, but it is a NIT.

If you strongly disagree, I want push for it.

File: smt.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply smt.go]: Let's add a check. If it errors return an error that says "This can only happen if a tree is empty".

It'll be a good way to catch errors / edge cases we may not be aware of (which should THEORETICALLY not happen)

File: smst_proofs_test.go (hunk 23 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply smst_proofs_test.go:179]: Upload an image in one of the comments and use that URL

File: smt.go [hunk 25]
```diff
@@ -398,6 +398,9 @@ func (smt *SMT) ProveClosest(path []byte) (
 	var siblings []treeNode
 	var sib treeNode
 	var parent treeNode
+	// depthDelta is used to track the depth increase when traversing down the tree
```
Comment (Olshansk) [inline smt.go:401]: ty!

File: smt.go (hunk 4 above)
Comment (Olshansk) [reply smt.go:429]: Please add that explanation in the comments. Very valuable information!

File: smt.go [hunk 26]
```diff
@@ -451,7 +460,10 @@ func (smt *SMT) ProveClosest(path []byte) (
 				node = ext.expand()
 			}
 		}
-		inner := node.(*innerNode)
+		inner, ok := node.(*innerNode)
+		if !ok { // this can only happen for an empty tree
+			break
```
Comment (Olshansk) [inline smt.go:465]: add a print statement!

File: smt.go [hunk 27]
```diff
@@ -463,10 +475,14 @@ func (smt *SMT) ProveClosest(path []byte) (
 	}
 
 	// Retrieve the closest path and value hash if found
-	if node == nil {
-		panic("no leaf node found")
+	if node == nil { // tree was empty
```
Comment (Olshansk) [inline smt.go:478]: I think we should print that the tree was empty.

If/when we hit bugs, this is the sort of thing that helps find it.

File: smt.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply smt.go:498]: Add a TODO with an exaplanation

File: smt_proofs_test.go [hunk 28]
```diff
@@ -156,6 +156,8 @@ func TestSMT_ProofsSanityCheck(t *testing.T) {
 	require.NoError(t, smv.Stop())
 }
 
+// ProveClosest test against a visual representation of the tree
+// See: https://i.imgur.com/ux4fchQ.png
```
Comment (Olshansk) [inline smt_proofs_test.go:160]: Use the github URL?

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: smt.go (hunk 27 above)
Comment (h5law) [reply smt.go:478]: I personally would be against including print statements in the library itself only included when doing debugging otherwise they would just bloat out the SMT during normal use. I think having added the new return specific to empty trees allows us to determine the outcome without needing the print statement:
1. Was the `closestPath` returned a placeholder & is the root hash a placeholder & was `closestValueHash` returned nil?
  - Tree 100% empty
2. Otherwise tree is non-empty

File: smt.go (hunk 26 above)
Comment (h5law) [reply smt.go:467]: @Olshansk Mentioned this below but I think the result of the return values should be used instead of a print. Besides debugging the SMT itself is there a usecase for print statements with the new return values clearly showing the tree was empty vs. 1 node vs N nodes

**Comment by Olshansk** [comment]
@h5law Seems like this is out of date with the main branch and might need to be rebased or merged.

File: smst_proofs_test.go (hunk 13 above)
Comment (Olshansk) [reply smst_proofs_test.go:208]: Let's update the 4 line comment to this:
```
// `testKey2` is the child of an inner node, which is the child of an extension node.
// The extension node has the path bounds of [3, 7]. This means any bits between
// 3-6 can be flipped, and the resulting path would still traverse through the same 
// extension node and lead to testKey2 - the the closest key. However, flipping bit 7
// will lead to testKey4.
```

**Comment by Olshansk** [comment]
> Look into removing SiblingData from proofs
> Unexport GetPathBit

Leave TODOs next to the structure/function with next steps. It's the most lightweight way to keep track of work (no tickets, etc) without losing context if both of us are not around.

File: smt.go (hunk 26 above)
Comment (Olshansk) [reply smt.go:467]: Fair enough. Resolving

**Review by Olshansk — DISMISSED** [review]
:shipit:
